// @parrotfi/msig/lib/src/instructions/ProposalBase.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProposalBase = void 0;
const web3_js_1 = require("@solana/web3.js");
const tweetnacl_1 = __importDefault(require("tweetnacl"));
const base64_js_1 = require("base64-js");
const utils_1 = require("../utils");
class ProposalBase {
    memo;
    accounts;
    constructor(memo, accounts) {
        this.memo = memo;
        this.accounts = accounts;
    }
    calcAccountWithSeed(seed) {
        return web3_js_1.Keypair.fromSeed(tweetnacl_1.default.hash(Uint8Array.from(Buffer.from(seed))).slice(0, 32));
    }
    calcTransactionAccount() {
        const seed = tweetnacl_1.default.hash(Uint8Array.from(Buffer.from(this.memo)));
        return web3_js_1.Keypair.fromSeed(seed.slice(0, 32));
    }
    // throw error if verify failed
    async verifyTx(ctx, chainTxState, verbose) {
        const instrs = await this.createInstr(ctx);
        const multisigInstr = instrs.multisigInstr;
        if (multisigInstr.keys.length != chainTxState.accounts.length) {
            throw Error(`verify failed, accounts length not match, local: ${multisigInstr.keys.length}, chain: ${chainTxState.accounts.length}`);
        }
        for (let i = 0; i < multisigInstr.keys.length; i++) {
            if (!multisigInstr.keys[i].pubkey.equals(chainTxState.accounts[i].pubkey)) {
                throw Error(`verify failed, accounts (index: ${i}) not match`);
            }
        }
        if (!multisigInstr.data.equals(chainTxState.data) ||
            !multisigInstr.programId.equals(chainTxState.programId)) {
            throw Error("verify failed, programId or instruction data not match");
        }
        if (verbose) {
            console.log("multisig instruction:");
            console.log("program id:", multisigInstr.programId.toBase58());
            console.log("accounts:");
            (0, utils_1.printKeys)(multisigInstr.keys);
            console.log("local created instr in base64: ", (0, base64_js_1.fromByteArray)(multisigInstr.data));
        }
    }
}
exports.ProposalBase = ProposalBase;
